CenterForOpenScience/osf.io
# [PLAT-1056] "Impact" preprint metrics back end

## Purpose



Provide the back end for the Impact feature. Epic: https://openscience.atlassian.net/browse/PRODUCT-530

## Changes


* Adds django-elasticsearch-metrics as a dependency
* Adds elasticsearch2 as a dependency (used for OSF search)
* Metrics are defined in `osf/metrics.py`
* Preprint metrics are recorded in the  waterbutler auth callback (`get_auth`)
* In APIv2, metrics are enabled by subclassing `api.base.metrics.MetricMixin` for views and `api.base.metrics.MetricsSerializerMixin` for serializers

## QA Notes



This change is back end only. This feature should get QA'd when the frontend goes out. In the meantime, I'll do some spot-checking on staging and prod (when this goes out) to make sure that metrics are making it to elasticsearch.


## Documentation


This feature is considered unstable and will not be added to our public API documentation.

I've added internal documentation here: https://www.notion.so/cos/Impact-API-documentation-6d7c638c0cb642f8989287a1794580b2

Documentation for django-elasticsearch-metrics is in the README: https://github.com/sloria/django-elasticsearch-metrics

## Side Effects


* In order for this feature to work for local dev, you need to start the elasticsearch6 container, enable the elasticsearch_metrics waffle switch, and run `sync_metrics`

```
docker-compose up -d elasticsearch6
python manage.py waffle_switch elasticsearch_metrics on --create
python manage.py sync_metrics
```

* **For release to prod**:

- [x] Elasticsearch 6 cluster deployed
- [x] `ELASTICSEARCH_DSL` setting must be configured in `api/base/settings/local.py`
- [ ] In Jenkins: `HELM_PROD_OSF` needs to be updated to 0.15.4. This prevents pods from updating before `migrate` and `sync_metrics` have been run
- [ ] Run `python manage.py sync_metrics`
- [ ] Enable the `elasticsearch_metrics` waffle switch in admin

* Whenever a change is made to the mapping or settings for a metric, the release manager must run

```python
python manage.py sync_metrics
```

## Ticket


https://openscience.atlassian.net/browse/PLAT-1056
